### PR TITLE
feat(gossipsub): make gossipsub modules private

### LINKED ISSUE #3494: refactor `gossipsub`: make modules private
## Description

As described [in this comment](https://github.com/libp2p/rust-libp2p/pull/3303#discussion_r1084698192):
- [ ] `protocol` - [PR](https://github.com/libp2p/rust-libp2p/pull/3495)
- [ ] `metrics` - [PR](https://github.com/libp2p/rust-libp2p/pull/3565)
- [ ] `subscription_filter` - [PR](https://github.com/libp2p/rust-libp2p/pull/3566)
- [ ] `time_cache` - [PR](https://github.com/libp2p/rust-libp2p/pull/3564)

## Are you planning to do it yourself in a pull request?

Yes

### PR DESCRIPTION
**Pull request by tcoratger**
## Description

Resolves #3494.

## Change checklist

- [x] I have performed a self-review of my own code
- [x] A changelog entry has been made in the appropriate crates

### CODE REVIEW

**Review by thomaseizinger — COMMENTED** [review]
From memory, I think at least some types of the `metrics` module appear in the public API of `gossipsub::Behaviour`.

Can you spot check that and re-export them from the root crate?

**Comment by tcoratger** [comment]
> From memory, I think at least some types of the `metrics` module appear in the public API of `gossipsub::Behaviour`.
> 
> Can you spot check that and re-export them from the root crate?

Done

**Comment by thomaseizinger** [comment]
I pushed some follow-up commits. Should be good to merge now!

File: protocols/gossipsub/src/subscription_filter_priv.rs [hunk 1]
```diff
@@ -200,62 +200,20 @@ where
...
+    fn can_subscribe(&mut self, topic_hash: &TopicHash) -> bool {
+        self.0.is_match(topic_hash.as_str())
     }
 }
 
 #[cfg(test)]
 mod test {
     use super::*;
     use crate::types::SubscriptionAction::*;
+    use regex::Regex;
```
Comment (thomaseizinger) [inline protocols/gossipsub/src/subscription_filter_priv.rs]:
```suggestion
```

**Comment by tcoratger** [comment]
> I pushed some follow-up commits. Should be good to merge now!

Yes I have just merged from `master` as there was conflicts but then should be good for merging, what do you think?

**Review by thomaseizinger — DISMISSED** [review]
Thanks!

**Pull request merged**